mypyc/irbuild/expression.py

```diff
@@ -436,23 +436,36 @@ def transform_comparison_expr(builder: IRBuilder, e: ComparisonExpr) -> Value:
             and isinstance(e.operands[1], (TupleExpr, ListExpr))):
         items = e.operands[1].items
         n_items = len(items)
-        # x in y -> x == y[0] or ... or x == y[n]
-        # x not in y -> x != y[0] and ... and x != y[n]
+        # x in y -> (x is y[0] or x == y[0]) or ... or (x is y[n] or x == y[n])
+        # x not in y -> (x is not y[0] and x != y[0]) and ... and (x is not y[n] and x != y[n])
         # 16 is arbitrarily chosen to limit code size
-        if 1 < n_items < 16:
+        if 1 <= n_items < 16:
+            lhs = e.operands[0]
             if e.operators[0] == 'in':
+                # check `is` relation statically
```

Comment:
We shouldn't use the `is` check for any unboxed types, such as `int_rprimitive`, since it's mostly redundant and can cause unnecessary boxing. For example, look at the type of the operands, and use the extra `is` check only if both of the operands are boxed types (use `is_unboxed` of `RType`).

As an optional optimization, if both types are straightforward native classes, and we know all the subclasses, and there is no custom `__eq__` in any of the classes or subclasses, we know that there is no `__eq__` override. This means that the `==` check is the same as an `is` check, and we can skip the latter one.

Reply:
Sorry to respond late. Thank you for your review and I commit to solve the first problem. As for the optional optimization, I can't figure out what do you mean by straightforward native classes and how can we know all the subclasses.